src/grammar.adoc

```diff
@@ -0,0 +1,208 @@
+                \|  `(` `*all_from*` _module-name_ `)`
+|===
+
+
+=== Macro Definitions
+
+[{bnf}]
+|===
+|macro-defn |::=|  `(` `*macro*` _macro-name_? _signature_ _template_ `)`
+
+|signature   |::=|  _param-specs_ _result-spec_?
+|param-specs |::=|  `(` _param-spec_* _rest-spec_? `)`  \|  `[` _param-spec_* _rest-spec_? `]`
+|param-spec  |::=|  _param-name_  \|  _simple-param-spec_  \|  _grouped-param-spec_
+|param-name  |::=|  _unannotated-identifier-symbol_
+
+|simple-param-spec  |::=|  `(` _param-name_  _base-shape_ _simple-cardinality_ `)`
+|base-shape         |::=| _tagged-type_  \|  _primitive-type_  \|  _macro-ref_
+|simple-cardinality |::=|  `*!*`  \|  `*+*`  \|  `*'?'*`  \|  `*'{asterisk}'*`
+
+|grouped-param-spec  |::=|  `(` _param-name_ `[` _base-shape_ `]` _grouped-cardinality_? `)`
+|grouped-cardinality |::=|  `*+*`  \|  `*'{asterisk}'*`
+
+|rest-spec        |::=|  `(` _param-name_ _base-shape_ _rest-cardinality_  `)`
+|rest-cardinality |::=|   `*\...*`  \|  `*\...+*`
+
+|tagged-type   |::=| _abstract-type_ \| _concrete-type_
+|abstract-type |::=| `*any*` \| `*number*` \| `*exact*` \| `*text*` \| `*lob*` \| `*sequence*`
+|concrete-type |::=|  `*'null'*`  \|  `*bool*`  \|  `*timestamp*`  \|  `*int*`  \|  `*decimal*`
+                  \|  `*float*`  \|  `*string*`  \|  `*symbol*`  \|  `*blob*`  \|  `*clob*`
+                  \|  `*list*`  \|  `*sexp*`  \|  `*struct*`
+
+|primitive-type |::=|  `*var_symbol*`  \|  `*var_string*`
+                   \|  `*var_int*`  \|  `*var_uint*`
+                   \|  `*uint8*`    \|  `*uint16*`   \|  `*uint32*`  \|  `*uint64*`
+                   \|  `*int8*`     \|  `*int16*`    \|  `*int32*`   \|  `*int64*`
+                   \|  `*float16*`  \|  `*float32*`  \|  `*float64*`
+
+|result-spec |::=|  `*\->*` _tagged-type_ _simple-cardinality_
+|===
+
+
+=== Template Expression Language
+
+[{bnf}]
+|===
+|template |::=|  _identifier_  \|  _literal_  \|  _quasi-literal_
+             \|  _special-form_  \|  _macro-invocation_
+
+|literal       |::=|  _null_  \|  _bool_  \|  _int_  \|  _float_  \|  _decimal_  \| _timestamp_
+                  \|  _string_  \|  _blob_  \|  _clob_
+
+|quasi-literal |::=|  `[` _template_* `]`  \|  `{` _quasi-field_* `}`
+|quasi-field   |::=| _text_ `:` _template_
+
+|special-form |::=|  `(` `*literal*` _datum_ `)`                             +
+                 \|  `(` `*if_void*`   _template_ _template_ _template_ `)`  +
+                 \|  `(` `*if_single*` _template_ _template_ _template_ `)`  +
+                 \|  `(` `*if_many*`   _template_ _template_ _template_ `)`  +
+                 \|  `(` `*when*`   _param-name_ _template_ `)`              +
+                 \|  `(` `*unless*` _param-name_ _template_ `)`              +
+                 \|  `(` `*for*` `[` _for-clause_* `]` _template_ `)`
+
+|for-clause       |::=| `(` _identifier_ _template_ `)`
+
+|macro-invocation |::=|  `(` _macro-ref_ _macro-arg_* `)`
```

Comment:
Technically this conflicts with the special forms in the grammar (since generally BNFs are not ordered, while PEGs are), probably worth noting that *special-form* takes precedence and one needs to use qualification to invoke a macro that shadows a special form within a template.

Reply:
Adding an admonition regarding that.